base/linalg/triangular.jl

```diff
@@ -283,6 +283,155 @@ function -(A::UnitUpperTriangular)
     UpperTriangular(Anew)
 end
 
+# copy and scale
+function copy!(A::UpperTriangular,B::UpperTriangular)
+    copy!(A.data,B.data)
+end
+function copy!(A::UnitUpperTriangular,B::UnitUpperTriangular)
+    copy!(A.data,B.data)
+end
+function copy!(A::LowerTriangular,B::LowerTriangular)
+    copy!(A.data,B.data)
+end
+function copy!(A::UnitLowerTriangular,B::UnitLowerTriangular)
+    copy!(A.data,B.data)
+end
+function copy(A::Union{UpperTriangular,UnitUpperTriangular,LowerTriangular,UnitLowerTriangular})
+    D = similar(A)
+    copy!(D, A)
+    return D
+end
+
+function scale!(A::UpperTriangular, B::UpperTriangular, c::Number)
+    n = chksquare(B)
+    for j = 1:n
+        for i = 1:j
+            @inbounds A[i,j] = c * B[i,j]
+        end
+    end
+end
+function scale!(A::UpperTriangular, B::UnitUpperTriangular, c::Number)
+    n = chksquare(B)
+    for j = 1:n
+        A[j,j] = c
+        for i = 1:j-1
+            @inbounds A[i,j] = c * B[i,j]
+        end
+    end
+end
+function scale!(A::LowerTriangular, B::LowerTriangular, c::Number)
+    n = chksquare(B)
+    for j = 1:n
+        A[j,j] = c
+        for i = j:n
+            @inbounds A[i,j] = c * B[i,j]
+        end
+    end
+end
+function scale!(A::LowerTriangular, B::UnitLowerTriangular, c::Number)
+    n = chksquare(B)
+    for j = 1:n
+        A[j,j] = c
+        for i = j+1:n
+            @inbounds A[i,j] = c * B[i,j]
+        end
+    end
+end
+
+function scale{R<:Real,C<:Complex}(A::Union{UpperTriangular{R}, LowerTriangular{R}}, c::C)
+    D = similar(A, promote_type(R,C))
+    scale!(D,A,c)
+    return D
+end
+function scale{R<:BigFloat,C<:Complex}(A::UpperTriangular{R}, c::C)
+    D = UpperTriangular(zeros(promote_type(R,C),size(A)...))
+    scale!(D,A,c)
+    return D
+end
+function scale{R<:BigFloat,C<:Complex}(A::LowerTriangular{R}, c::C)
+    D = LowerTriangular(zeros(promote_type(R,C),size(A)...))
+    scale!(D,A,c)
+    return D
+end
+function scale{R<:BigFloat,C<:Complex}(A::UnitUpperTriangular{R}, c::C)
+    D = UpperTriangular(zeros(promote_type(R,C),size(A)...))
+    scale!(D,A,c)
+    return D
+end
+function scale{R<:BigFloat,C<:Complex}(A::UnitLowerTriangular{R}, c::C)
+    D = LowerTriangular(zeros(promote_type(R,C),size(A)...))
+    scale!(D,A,c)
+    return D
+end
+function scale{C<:Complex}(A::UpperTriangular{BigFloat}, c::C)
+    D = UpperTriangular(zeros(promote_type(BigFloat,C),size(A)...))
+    scale!(D,A,c)
+    return D
+end
+function scale{C<:Complex}(A::LowerTriangular{BigFloat}, c::C)
+    D = LowerTriangular(zeros(promote_type(BigFloat,C),size(A)...))
+    scale!(D,A,c)
+    return D
+end
```

Comment:
In general, if you find yourself copying and pasting multiple versions of nearly identical code, you are making a mistake.

If the methods can't be unified by being more clever about the type signature, you should fall back to a loop of `@eval` invocations.